dart-mode.el

```diff
@@ -735,19 +745,33 @@ the callback for that request is given the json decoded response."
                                 (-butlast buf-lines)))))
             (-each json-lines 'dart--analysis-server-handle-msg)))))))
 
+(defun dart--execute-analysis-callback (msg id callbacks)
+  "Execute different callbacks depending on the kind of response received.
+
+Argument MSG is the parsed response from the analysis server.
+Argument ID is the id of the event or response sent by the analysis server.
+Argument CALLBACKS is the list containing the callback to be executed."
+  (-if-let (resp-closure (assoc id callbacks))
+      (progn
+	(setq callbacks
+	      (assq-delete-all id callbacks))
+	(funcall (cdr resp-closure) msg))
+    (-if-let (err (assoc 'error msg))
+	(dart--analysis-server-on-error-callback msg)
+      (dart-info (format "No callback was associated with id %s" id)))))
+
 (defun dart--analysis-server-handle-msg (msg)
   "Handle the parsed MSG from the analysis server."
+  (-when-let* ((event-assoc (assoc 'event msg))
```

Comment:
Using two `-when-let*`s in a row here is confusing, because it looks like a message could be both a reply and an event. I'd use `if` to choose between the two `-when-let*`s.